src/generic/complex_matrices.cc

```diff
@@ -538,6 +539,647 @@ namespace oomph
+      unsigned long M = matrix_in.ncol();
+      unsigned long Nnz = 0;
+
+      // pointers to arrays which store result
+      int* Row_start = 0;
+      std::complex<double>* Value = 0;
+      int* Column_index = 0;
+
+      // get pointers to matrix_in
+      const int* matrix_in_row_start = matrix_in.row_start();
+      const int* matrix_in_column_index = matrix_in.column_index();
+      const std::complex<double>* matrix_in_value = matrix_in.value();
+
+      // get pointers to this matrix
+      const std::complex<double>* this_value = this->value();
+      const int* this_row_start = this->row_start();
+      const int* this_column_index = this->column_index();
+
+      // clock_t clock1 = clock();
+
+      // METHOD 1
+      // --------
+      if (method == 1)
+      {
+        // allocate storage for row starts
+        Row_start = new int[N + 1];
+        Row_start[0] = 0;
+
+        // a set to store number of non-zero columns in each row of result
+        std::set<unsigned> columns;
+
+        // run through rows of this matrix and matrix_in to find number of
+        // non-zero entries in each row of result
+        for (unsigned long this_row = 0; this_row < N; this_row++)
+        {
+          // run through non-zeros in this_row of this matrix
+          for (int this_ptr = this_row_start[this_row];
+               this_ptr < this_row_start[this_row + 1];
+               this_ptr++)
+          {
+            // find column index for non-zero
+            int matrix_in_row = this_column_index[this_ptr];
+
+            // run through corresponding row in matrix_in
+            for (int matrix_in_ptr = matrix_in_row_start[matrix_in_row];
+                 matrix_in_ptr < matrix_in_row_start[matrix_in_row + 1];
+                 matrix_in_ptr++)
+            {
+              // find column index for non-zero in matrix_in and store in
+              // columns
+              columns.insert(matrix_in_column_index[matrix_in_ptr]);
+            }
+          }
+          // update Row_start
+          Row_start[this_row + 1] = Row_start[this_row] + columns.size();
+
+          // wipe values in columns
+          columns.clear();
+        }
+
+        // set Nnz
+        Nnz = Row_start[N];
+
+        // allocate arrays for result
+        Value = new std::complex<double>[Nnz];
+        Column_index = new int[Nnz];
+
+        // set all values of Column_index to -1
+        for (unsigned long i = 0; i < Nnz; i++)
+        {
+          Column_index[i] = -1;
+        }
+
+        // Calculate values for result - first run through rows of this matrix
+        for (unsigned long this_row = 0; this_row < N; this_row++)
+        {
+          // run through non-zeros in this_row
+          for (int this_ptr = this_row_start[this_row];
+               this_ptr < this_row_start[this_row + 1];
+               this_ptr++)
+          {
+            // find value of non-zero
+            std::complex<double> this_val = this_value[this_ptr];
```

Comment:
Same here with whatever is decided for `this_`-style naming

Reply:
See above